### PR TITLE
Add files with the API for all fields, binary fields, and prime extension fields

### PR DESCRIPTION
**Pull request by alexander-zw**
Also changed .gitignore and the README

### CODE REVIEW

File: libff/algebra/fields/binary_field.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /* Functions common to all finite fields */
+#ifdef PROFILE_OP_COUNTS // NOTE: op counts are affected when you exponentiate with ^
+    static long long add_cnt;
+    static long long sub_cnt;
+    static long long mul_cnt;
+    static long long sqr_cnt;
+    static long long inv_cnt;
+#endif
+
+    virtual T& operator+=(const T& other) = 0;
```
Comment (ValarDragon) [inline libff/algebra/fields/binary_field.hpp:57]: Is it okay for these to be virtual, will we lose efficiency due to vtable issues?

File: libff/algebra/fields/field.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    void randomize() = 0;
+    void clear() = 0;
+
+    // the following should be defined in child classes, but are static so they can't be inherited
+    static T zero();
+    static T one();
+    static T random_element();
+    /** Equals 1 for prime field Fp. */
+    static constexpr std::size_t extension_degree();
+    static std::size_t size_in_bits();
```
Comment (ValarDragon) [inline libff/algebra/fields/field.hpp:81]: If a modulus is `2^n + 1` (or any n+1 bit number), does this return `n` or `n+1`?

File: README.md [hunk 3]
```diff
@@ -82,6 +86,11 @@ Optionally, you can specify the install location by providing the desired instal
 ```
 cmake .. -DCMAKE_INSTALL_PREFIX=/install/path
 ```
+On Mac OS X, you may also need to turn off `libprocps` and/or provide the path to openssl (your exact path may vary):
```
Comment (ValarDragon) [inline README.md:89]: We've now changed the default flag to have WITH_PROCPS=off, so I think this part of the change can be removed!

**Review by ValarDragon — COMMENTED** [review]
Thanks for the work on this, and splitting up the PR!

The field API looks generally good to me, left some contents relating to some more minor changes.

Just to check, theres no difference between the binary field and prime field API now other than the `as_words` right?

File: libff/algebra/fields/binary_field.hpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class gf32 : public BinaryField<gf32> ...
+ */
+template<typename T>
+class BinaryField {
+public:
+    /* Functions unique to binary fields */
+
+    /** Returns the constituent bits in 64 bit words, in little-endian order. */
```
Comment (AntoineRondelet) [inline libff/algebra/fields/binary_field.hpp:35]: Is there a reason why you use comments of the form `/* <my comment> */` and of the form `/** <my comment> */`?
If not, that's inconsistent and thus I'd favor one style everywhere.
If the goal is to use Doxygen, we need to stick to one supported code block style: https://www.doxygen.nl/manual/docblocks.html

Hence:
```
/**
 * <my comment>
 */
```
Would work best here

File: libff/algebra/fields/field.hpp (hunk 2 above)
Comment (AntoineRondelet) [reply libff/algebra/fields/field.hpp:81]: IIRC, `size_in_bits(x)` should return `ceil(log_2(x))` (i.e. `n+1` in your example above) (`capacity(x)`, however, returns `floor(log_2(x))` (i.e. `n` in your eg))

Comment (ValarDragon) [reply libff/algebra/fields/field.hpp:81]: That makes sense. Should we instead call them `floor_size_in_bits` and `ceil_size_in_bits` to remove ambiguity?

File: libff/algebra/fields/binary_field.hpp (hunk 4 above)
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:35]: /** */ is used for documentation, while /* */ is used to mark a new section of code, and // is general comments that explain some code, but is not high level enough to be included in documenation.

Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:35]: The documentation `/** */` when it's one line and
```
/**
 *
 */
```
otherwise

File: libff/algebra/fields/binary_field.hpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    std::vector<uint64_t> as_words() const = 0;
```
[the remark this answers is not in the document]
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:36]: Regarding from_words(), this would probably not be too hard, but why would it be needed? There are already the constructor, which takes in uint64_t as arguments, and the >> operator.

[the remark this answers is not in the document]
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:36]: For adding this to Fp^n, how would that be done? Fp has as_bigint() and as_ulong(), but it doesn't make as much sense to convert to a vector<uint64_t> for Fp or Fp^n. If it is needed to store the value in a string there is << and >>.

File: libff/algebra/fields/binary_field.hpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    void randomize() = 0;
+    void clear() = 0;
+
+    // the following should be defined in child classes, but are static so they can't be inherited
+    static T zero();
+    static T one();
+    static T random_element();
+    static constexpr std::size_t extension_degree();
+    static std::size_t size_in_bits() { return num_bits; }
```
[the remark this answers is not in the document]
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:94]: Originally I had the default implementation with the intention of having classes inherit it, but as you know that didn't happen. So I included it here as essentially documentation.

File: libff/algebra/fields/field.hpp (hunk 2 above)
Comment (alexander-zw) [reply libff/algebra/fields/field.hpp:81]: I think adding a comment explaining should be fine?

File: README.md (hunk 3 above)
Comment (alexander-zw) [reply README.md:89]: In that case we can clarify what this flag does somewhere

**Comment by alexander-zw** [comment]
> Just to check, theres no difference between the binary field and prime field API now other than the `as_words` right?

Aren't there the public members and functions like modulus_, multiplicative_generator, euler, Frobenius_map, etc?

File: libff/algebra/fields/binary_field.hpp (hunk 4 above)
Comment (AntoineRondelet) [reply libff/algebra/fields/binary_field.hpp:35]: Oh right, didn't know `/**<comment> */` was also a valid Doxygen syntax. Sounds good!

File: README.md (hunk 3 above)
Comment (ValarDragon) [reply README.md:89]: Good idea. Perhaps we should have a dedicated INSTALL.md file that details setting up the library if your not on mac os, and all the optional flags.

File: libff/algebra/fields/binary_field.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply libff/algebra/fields/binary_field.hpp:36]: In libiop we need to construct field elements from raw byte strings. Previously we did this by using the data field directly, which was a hack. I guess we could support it by implementing a from_bytes method?

From_words for Fp isn't so weird, there is a notion of words within bigint. (In particular, the bigint.data field)

It is a bit weird in Fp^n, but we could repeat the idea for Fp, `n` times.

File: libff/algebra/fields/field.hpp (hunk 2 above)
Comment (ValarDragon) [reply libff/algebra/fields/field.hpp:81]: hrmm, I'd strongly prefer an unambiguous function name. This feels like it can cause confusion / off-by-one soundness errors to downstream users.

Comment (alexander-zw) [reply libff/algebra/fields/field.hpp:81]: Right now only Fp implements capacity(), I'll change it to make all fields implement it, renamed floor_size_in_bits()

File: libff/algebra/fields/binary_field.hpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    // the following should be defined in child classes, but are static so they can't be inherited
+    static T zero();
+    static T one();
+    static T random_element();
+    /**
+     * Creates a field element from the given bits in 64 bit words, in little-endian order.
+     * Only the right-most size_in_bits() bits are used; other bits are ignored.
+     */
+    static T from_words(std::vector<uint64_t> words);
```
Comment (ValarDragon) [inline libff/algebra/fields/binary_field.hpp:100]: can you move this to be next to `as_words`, and perhaps rename `as_words` to `to_words` for name consistency?

File: libff/algebra/fields/binary_field.hpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class gf32 : public BinaryField<gf32> ...
+ */
+template<typename T>
+class BinaryField {
+public:
+    /* Functions unique to binary fields */
+
+    static const constexpr uint64_t modulus_;
```
Comment (ValarDragon) [inline libff/algebra/fields/binary_field.hpp:35]: TODO: We should eventaully add some documentation about how modulii are represented, no need to block this PR though!

File: libff/algebra/fields/binary_field.hpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+ *****************************************************************************
+ Declaration of common API for all finite fields in the binary/ directory.
+
+ Currently NOT used by the fields in this library. This class is not actually
+ the parent class of any field. All APIs are enforced through tests instead.
+
+ The reason for this is to ensure high performance of all fields. This class
+ exists as documentation for common API between fields.
+
+ Includes fields F2^n for specified n. All of the binary entension fields must
```
Comment (ValarDragon) [inline libff/algebra/fields/binary_field.hpp:11]: Can F2^n be changed to GF(2^n) or F_{2^n}?

File: libff/algebra/fields/field.hpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+ *****************************************************************************
+ Declaration of common API for all finite fields.
+
+ Currently NOT used by the fields in this library. This class is not actually
+ the parent class of any field. All APIs are enforced through tests instead.
+
+ The reason for this is to ensure high performance of all fields. This class
+ exists as documentation for common API between fields.
+
+ Includes two types of fields, F[p^n] for selected n and F[2^n] for a separate
```
Comment (ValarDragon) [inline libff/algebra/fields/field.hpp:11]: Nice! Didn't realize this API would support both prime and binary field extensions

**Review by ValarDragon — COMMENTED** [review]
Thanks for the fixes, a couple of minor comments

File: README.md [hunk 11]
```diff
@@ -78,11 +82,18 @@ To compile, starting at the project root directory, create the build directory a
...
 ```
-Optionally, you can specify the install location by providing the desired install path prefix:
+On MacOS, you may also need to provide the path to openssl (your exact path may vary):
 ```
-cmake .. -DCMAKE_INSTALL_PREFIX=/install/path
+cmake -DOPENSSL_ROOT_DIR=/usr/local/Cellar/openssl/1.0.2s -DOPENSSL_LIBRARIES=/usr/local/Cellar/openssl/1.0.2s/lib ..
 ```
 
+Other build flags include:
+| Flag | Value | Description |
```
Comment (ValarDragon) [inline README.md:91]: Thanks for starting on flags! Can we actually do these README updates in a separate PR? (I think this way of updating the SSL root directory is more likely to break than whats on staging atm)

File: libff/algebra/fields/binary_field.hpp (hunk 7 above)
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:100]: I put it here because I was placing static functions together. Do you think it's better to move it?

File: README.md (hunk 11 above)
Comment (alexander-zw) [reply README.md:91]: I'm not sure what you mean. Do you want me to separate the README changes to a different PR? Or do you want to add more README changes than the ones in this PR?

File: libff/algebra/fields/binary_field.hpp (hunk 7 above)
Comment (ValarDragon) [reply libff/algebra/fields/binary_field.hpp:100]: Yeah I think its better to move it

File: README.md (hunk 11 above)
Comment (ValarDragon) [reply README.md:91]: Yeah I mean lets separate the README changes into a separate PR

File: libff/algebra/fields/binary_field.hpp (hunk 7 above)
Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:100]: Hmm I feel like I disagree, but eh it's not a big deal

Comment (alexander-zw) [reply libff/algebra/fields/binary_field.hpp:100]: I decided to move as_words (now to_words) down, instead of moving to_words up

File: README.md (hunk 11 above)
Comment (alexander-zw) [reply README.md:91]: I can still keep the .gitignore changes right?
